Mathlib/GroupTheory/Perm/Cycle/Factors.lean

```diff
@@ -519,6 +529,40 @@ lemma support_zpowers_of_mem_cycleFactorsFinset_le {g : Perm α}
   simp only [← hm]
   exact le_trans (support_zpow_le _ _) (mem_cycleFactorsFinset_support_le c.prop)
 
+theorem pairwise_disjoint_of_mem_zpowers :
+    Pairwise fun (i j : f.cycleFactorsFinset) ↦
+      ∀ (x y : Perm α), x ∈ Subgroup.zpowers ↑i → y ∈ Subgroup.zpowers ↑j → Disjoint x y :=
+  fun c d  hcd ↦ fun x y hx hy ↦ by
+  obtain ⟨m, hm⟩ := hx; obtain ⟨n, hn⟩ := hy
+  simp only [← hm, ← hn]
+  apply Disjoint.zpow_disjoint_zpow
+  exact f.cycleFactorsFinset_pairwise_disjoint c.prop d.prop (Subtype.coe_ne_coe.mpr hcd)
+
+lemma pairwise_commute_of_mem_zpowers :
+    Pairwise fun (i j : f.cycleFactorsFinset) ↦
+      ∀ (x y : Perm α), x ∈ Subgroup.zpowers ↑i → y ∈ Subgroup.zpowers ↑j → Commute x y :=
+  f.pairwise_disjoint_of_mem_zpowers.mono
+    (fun _ _ ↦ forall₂_imp (fun _ _ h hx hy ↦ (h hx hy).commute))
+
+lemma disjoint_ofSubtype_noncommPiCoprod (u : Perm (Function.fixedPoints ⇑f))
+    (v : (c : { x // x ∈ f.cycleFactorsFinset }) → ↥(Subgroup.zpowers (c : Perm α))) :
+    Disjoint (ofSubtype u) ((Subgroup.noncommPiCoprod f.pairwise_commute_of_mem_zpowers) v) := by
+  apply Finset.noncommProd_induction
+  · intro a _ b _ h
+    apply f.pairwise_commute_of_mem_zpowers h <;> simp only [Subgroup.coeSubtype, SetLike.coe_mem]
+  · intro x y
+    exact Disjoint.mul_right
+  · exact disjoint_one_right _
+  · intro c _
+    simp only [Subgroup.coeSubtype]
+    exact Disjoint.mono (disjoint_ofSubtype_of_memFixedPoints_self u)
+      le_rfl (support_zpowers_of_mem_cycleFactorsFinset_le (v c))
+
+lemma commute_ofSubtype_noncommPiCoprod (u : Perm ↑(Function.fixedPoints ⇑f))
+    (v : (c : { x // x ∈ f.cycleFactorsFinset }) → ↥(Subgroup.zpowers (c : Perm α))) :
```

Comment:
```suggestion
lemma commute_ofSubtype_noncommPiCoprod (u : Perm (Function.fixedPoints f))
    (v : (c : { x // x ∈ f.cycleFactorsFinset }) → Subgroup.zpowers (c : Perm α)) :
```

I think not writing function coercion or coeSort arrows is better style

Reply:
OK.